### ISSUE TITLE
Adding link types to org-roam

### ISSUE THREAD

**Issue by Cletip** [org-roam/org-roam#2318]
### Brief Abstract

Need to add link types in org-roam, to create relationships of the form : note1 -> relationship -> note2/OrValue.

### Long Description

This issue has been discussed a lot, I know. I have browsed many forums, and no solution is available for org-roam V2.

So I propose to make an implementation of this, with my (poor) skills

### What I expect from this issue

A relevant opinion on the choice of the implementation to make: directly modify org-roam, make a separate package, which method to choose, etc.

### Proposed Implementation

Several solutions:

- create links like this: [[id1:id2][description]]
Here, id1 is the type of the relationship, id2 is the value of the relationship. id2 is misnamed: it can be an id, or just a simple value (a string or a number etc)
Advantages : no need to modify anything in org-roam database, quite easy to do in org-mode (org-element-link-parser and org-open-link) and org-roam (change the function that retrieves backlinks, I personally think I have all types of links in backlinks, and not only "id"), quite intuitive to use. Functional and tested on my configuration (I didn't test all of them, but I know that I would also have to modify the function to export to html, or things like that).

- or links like this: [[id-rel:id1::id2][description]]. Creating a new type of link: id-rel.
I don't see how to do this so as not to drastically change the database queries (each time you have to check the two types: id and id-rel), plus you have to separate id1::id2 each time, or else create a new column in the database, which would be "empty" if the link is of type "id", and "id1" if it is of type "id-rel". This would be the "relationship" column.

**Comment by crlsh** [comment]
related??
https://org-roam.discourse.group/t/link-categorization/2486

**Comment by Cletip** [comment]
Yes, largely, and I have already browsed these forums, especially this one :+1: https://org-roam.discourse.group/t/add-link-tags-feature/171.
Unfortunately, to my knowledge, nothing has been done concretely.

**Comment by czan** [comment]
A little while ago I wrote an implementation of link types: https://gist.github.com/czan/87f80b807be270344e7450fce6c9deb8 that is basically just [what Jethro suggested in June 2021](https://org-roam.discourse.group/t/add-link-tags-feature/171/34).

After loading that gist, use `M-x org-roam-relationship-mode` to enable the custom link behaviours.

This works by defining new Org link types for relationships, and prompting for a relationship when inserting links with `org-roam-node-insert`. You can use `customise` to add relationship types to `org-roam-relationship-types`, and the backlinks buffer should show separate sections for each link type. There's not much to it, so it shouldn't be too hard to change if you want it to behave differently.

I haven't actually found link types to be very useful, so I'm not looking to maintain this code, but if it's a useful starting point for someone else feel free to take whatever you want.
